### PR TITLE
Updating build pipelines to Cache@2 and switching to artifacts for compiled file sharing in pipeline jobs. 

### PR DESCRIPTION
**Pull request by aasimkhan30**

### CODE REVIEW

File: build/azure-pipelines/common/computeNodeModulesCacheKey.ts [hunk 1]
```diff
@@ -0,0 +1,51 @@
+/*---------------------------------------------------------------------------------------------
```
Comment (alanrenmsft) [inline build/azure-pipelines/common/computeNodeModulesCacheKey.ts:1]: how about adding a `sql-` prefix to the file name? make it easy to tell it is created by us

File: build/azure-pipelines/common/computeNodeModulesCacheKey.ts [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+  *  Copyright (c) Microsoft Corporation. All rights reserved.
+  *  Licensed under the MIT License. See License.txt in the project root for license information.
```
Comment (alanrenmsft) [inline build/azure-pipelines/common/computeNodeModulesCacheKey.ts:3]: the file header is incorrect, please copy the header from another file, it should be under the `SOURCE EULA`.

**Comment by alanrenmsft** [comment]
please create a few test runs to:
1. cover all the scenarios: cache miss, cache hit
2. produce a build and test it locally to verify ADS works as expected.

File: build/azure-pipelines/common/listNodeModules.ts [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+import * as path from 'path';
+
+if (process.argv.length !== 3) {
+	console.error('Usage: node listNodeModules.js OUTPUT_FILE');
+	process.exit(-1);
+}
+
+const ROOT = path.join(__dirname, '../../../');
+
+function findNodeModulesFiles(location: string, inNodeModules: boolean, result: string[]) {
```
Comment (Charles-Gagnon) [inline build/azure-pipelines/common/listNodeModules.ts:18]: Same, should be able to use glob here to make this simpler

File: build/azure-pipelines/darwin/sql-product-build-darwin.yml [hunk 4]
```diff
@@ -5,26 +5,18 @@ steps:
...
     inputs:
-      keyfile: 'build/.cachesalt, .build/commit, .build/quality'
-      targetfolder: '.build, out-build, out-vscode-min, out-vscode-reh-min, out-vscode-reh-web-min'
-      vstsFeed: 'BuildCache'
-      platformIndependent: true
-      alias: 'Compilation'
+      artifact: Compilation
+      path: $(Build.ArtifactStagingDirectory)
+    displayName: Download compilation output
+    condition: and(succeeded(), ne(variables['VSCODE_ARCH'], 'universal'))
```
Comment (Charles-Gagnon) [inline build/azure-pipelines/darwin/sql-product-build-darwin.yml:13]: Where is this new env var coming from?

File: build/azure-pipelines/common/computeNodeModulesCacheKey.js [hunk 5]
```diff
@@ -0,0 +1,43 @@
+/*---------------------------------------------------------------------------------------------
```
Comment (Charles-Gagnon) [inline build/azure-pipelines/common/computeNodeModulesCacheKey.js]: Looks like you're building this directory in the pipelines anyways, so any reason to have this checked in?

File: build/azure-pipelines/darwin/sql-product-build-darwin.yml [hunk 6]
```diff
@@ -61,34 +53,45 @@ steps:
       git merge $(node -p "require('./package.json').distro")
     displayName: Merge distro
 
-  - task: 1ESLighthouseEng.PipelineArtifactCaching.RestoreCacheV1.RestoreCache@1
+  - script: |
+      mkdir -p .build
+      node build/azure-pipelines/common/computeNodeModulesCacheKey.js > .build/yarnlockhash
+    displayName: Prepare yarn cache flags
```
Comment (Charles-Gagnon) [inline build/azure-pipelines/darwin/sql-product-build-darwin.yml:59]: flags? This isn't really a flag. `Prepare yarn cache hash` or `Prepare yarn cache key` seems more appropriate

File: build/azure-pipelines/linux/sql-product-build-linux.yml [hunk 7]
```diff
@@ -57,34 +47,45 @@ steps:
       git merge $(node -p "require('./package.json').distro")
     displayName: Merge distro
 
-  - task: 1ESLighthouseEng.PipelineArtifactCaching.RestoreCacheV1.RestoreCache@1
+  - script: |
+      mkdir -p .build
+      node build/azure-pipelines/common/computeNodeModulesCacheKey.js > .build/yarnlockhash
+    displayName: Prepare yarn cache flags
```
Comment (Charles-Gagnon) [inline build/azure-pipelines/linux/sql-product-build-linux.yml:53]: Same here about the name

File: build/azure-pipelines/sql-product-compile.yml [hunk 8]
```diff
@@ -102,15 +93,14 @@ steps:
...
-  condition: and(succeeded(), ne(variables['CacheRestored-Compilation'], 'true'))
 
 - script: |
     set -e
     AZURE_STORAGE_ACCOUNT="$(sourcemap-storage-account)" \
     AZURE_STORAGE_ACCESS_KEY="$(sourcemap-storage-key)" \
       node build/azure-pipelines/upload-sourcemaps
   displayName: Upload sourcemaps
-  condition: and(succeeded(), ne(variables['CacheExists-Compilation'], 'true'))
+  condition: and(succeeded(), ne(variables['VSCODE_PUBLISH'], 'false'))
```
Comment (Charles-Gagnon) [inline build/azure-pipelines/sql-product-compile.yml:103]: This seems unrelated, why was this updated?

File: build/azure-pipelines/sql-product-compile.yml [hunk 9]
```diff
@@ -122,15 +112,17 @@ steps:
     node build/azure-pipelines/common/copyArtifacts.js
   displayName: Write Version Information
 
-- task: PublishBuildArtifacts@1
-  displayName: 'Publish Artifact: drop'
+- script: |
+    set -e
+    tar -czf $(Build.ArtifactStagingDirectory)/compilation.tar.gz .build out-*
+  condition: and(succeeded(), ne(variables.COMPILATION_RESTORED, 'true'))
```
Comment (Charles-Gagnon) [inline build/azure-pipelines/sql-product-compile.yml:118]: COMPILATION_RESTORED doesn't exist anymore so you should be able to just get rid of this

**Review by Charles-Gagnon — CHANGES_REQUESTED** [review]

File: build/azure-pipelines/common/listNodeModules.ts (hunk 3 above)
Comment (aasimkhan30) [reply build/azure-pipelines/common/listNodeModules.ts:18]: I just copied what vscode was doing. Should I still look into using glob?

Comment (Charles-Gagnon) [reply build/azure-pipelines/common/listNodeModules.ts:18]: Eh, if you were copying what they were doing then this is fine I guess. But something to keep in mind in the future since it's a heck of a lot more readable.

File: build/azure-pipelines/sql-product-compile.yml (hunk 8 above)
Comment (aasimkhan30) [reply build/azure-pipelines/sql-product-compile.yml:103]: I just updated the flags for this task from vscode while removing CacheExists-Compilation check

File: build/azure-pipelines/darwin/sql-product-build-darwin.yml (hunk 6 above)
Comment (aasimkhan30) [reply build/azure-pipelines/darwin/sql-product-build-darwin.yml:59]: Using the task name from vscode

File: build/azure-pipelines/linux/sql-product-build-linux.yml (hunk 7 above)
Comment (aasimkhan30) [reply build/azure-pipelines/linux/sql-product-build-linux.yml:53]: Same. Got it from vscode

File: build/azure-pipelines/common/computeNodeModulesCacheKey.js (hunk 5 above)
Comment (aasimkhan30) [reply build/azure-pipelines/common/computeNodeModulesCacheKey.js]: I don't think I have built anything until this step. Atleast for the compile job.

File: build/azure-pipelines/darwin/sql-product-build-darwin.yml (hunk 4 above)
Comment (aasimkhan30) [reply build/azure-pipelines/darwin/sql-product-build-darwin.yml:13]: Got it from vscode. Will change as we don't support multi arch yet.

File: build/azure-pipelines/darwin/sql-product-build-darwin.yml (hunk 6 above)
Comment (Charles-Gagnon) [reply build/azure-pipelines/darwin/sql-product-build-darwin.yml:59]: Eh, since this is our file then I would still suggest updating them since - to me at least - these don't really give a good indication what they're actually doing.

File: build/azure-pipelines/common/computeNodeModulesCacheKey.ts [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+ const shasum = crypto.createHash('sha1');
+
+ shasum.update(fs.readFileSync(path.join(ROOT, 'build/.cachesalt')));
+ shasum.update(fs.readFileSync(path.join(ROOT, '.yarnrc')));
+ shasum.update(fs.readFileSync(path.join(ROOT, 'remote/.yarnrc')));
+
+
+ const result: string[] = [];
+ findFiles('', 'yarn.lock', result);
+ result.forEach(f => shasum.update(fs.readFileSync(f)));
```
[the remark this answers is not in the document]
Comment (aasimkhan30) [reply build/azure-pipelines/common/computeNodeModulesCacheKey.ts]: They are doing the same thing if I am not wrong. Check this file out. https://github.com/microsoft/vscode/blob/28a246d9ec5d10d9eaca7c80e1c942b1768d696b/build/azure-pipelines/common/computeNodeModulesCacheKey.ts#L23

[the remark this answers is not in the document]
Comment (aasimkhan30) [reply build/azure-pipelines/common/computeNodeModulesCacheKey.ts]: They have a hardcoded array of all the dirs that may contain yarn.lock files. Instead, I am just scanning the entire project for them. The worst that can happen is we might have few more cache misses because of scanning additional yarn.lock files. This doesn't seem harmful to me. 

Check this file out https://github.com/microsoft/vscode/blob/28a246d9ec5d10d9eaca7c80e1c942b1768d696b/build/npm/dirs.js

[the remark this answers is not in the document]
Comment (Charles-Gagnon) [reply build/azure-pipelines/common/computeNodeModulesCacheKey.ts]: Hm, would love to know why they did this. Unfortunately the changes made to this engineering stuff seems to be very light on the explanations of why things are being changed.

I agree with your approach though, if we start running into issues with that though we can look into adopting what they did but I'm guessing it's more related to the different architectures and modes that they compile for and improving speeds than any actual issues with doing it this way.

File: build/azure-pipelines/darwin/sql-product-build-darwin.yml [hunk 11]
```diff
@@ -61,34 +51,45 @@ steps:
...
-    displayName: Save Cache - Node Modules
-    inputs:
-      keyfile: 'build/.cachesalt, .yarnrc, remote/.yarnrc, **/yarn.lock, !**/node_modules/**/yarn.lock, !**/.*/**/yarn.lock, !samples/**/yarn.lock'
-      targetfolder: '**/node_modules, !**/node_modules/**/node_modules, !samples/**/node_modules'
-      vstsFeed: 'BuildCache'
-    condition: and(succeeded(), ne(variables['CacheRestored'], 'true'))
+  - script: |
+      set -e
+      node build/azure-pipelines/common/listNodeModules.js .build/node_modules_list.txt
+      mkdir -p .build/node_modules_cache
```
Comment (Charles-Gagnon) [inline build/azure-pipelines/darwin/sql-product-build-darwin.yml:83]: nit: spacing is slightly off here it looks like

File: build/azure-pipelines/common/sql-computeNodeModulesCacheKey.js [hunk 12]
```diff
@@ -0,0 +1,23 @@
+/*---------------------------------------------------------------------------------------------
```
Comment (Charles-Gagnon) [inline build/azure-pipelines/common/sql-computeNodeModulesCacheKey.js:1]: So currently we don't have these js files checked in for this folder for any of the other scripts - VS Code used to do this but it looks like they changed that a couple months back to improve compile times : 

https://github.com/microsoft/vscode/commit/d1fff1a08360f0d35f124c990dc65a6bccd748d2

But given that we don't do this for anything else currently I would say let's keep things consistent and not check in this one either since it's annoying to have to remember to update the compiled JS files for some of the scripts but not others.

File: build/azure-pipelines/common/listNodeModules.ts [hunk 13]
```diff
@@ -0,0 +1,46 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
```
Comment (Charles-Gagnon) [inline build/azure-pipelines/common/listNodeModules.ts:3]: This header is also incorrect

**Review by Charles-Gagnon — CHANGES_REQUESTED** [review]
Few smaller things, but overall looks good.

Have you done the testing that Alan suggested? Please verify this and provide links to the builds generated.

File: build/azure-pipelines/common/sql-computeNodeModulesCacheKey.js (hunk 12 above)
Comment (Charles-Gagnon) [reply build/azure-pipelines/common/sql-computeNodeModulesCacheKey.js:1]: Synced offline - doing this would require some additional work to add in steps for restoring packages and building the folder (this one is special since it has to run before the node_modules restore, so can't rely on that). So for now we'll keep it like this with the checked in JS and Aasim will be following up with changes to consolidate how we're handling the files here.

**Review by Charles-Gagnon — APPROVED** [review]
Signing off with the assumption that proper testing has been done with a successful build of this pipeline across all platforms.

**Comment by aasimkhan30** [comment]
@Charles-Gagnon, @alanrenmsft 

Initial cache miss: 
https://mssqltools.visualstudio.com/CrossPlatBuildScripts/_build/results?buildId=99841&view=results
Cache hit: 
https://mssqltools.visualstudio.com/CrossPlatBuildScripts/_build/results?buildId=99841&view=results
Cache miss after adding package: 
https://mssqltools.visualstudio.com/CrossPlatBuildScripts/_build/results?buildId=99875&view=results


I will just wait for the last build to finish and merge this pipeline. Please take a look and let me know if there is something wrong with the three builds.

**Comment by aasimkhan30** [comment]
Merging this in as the builds look good to me. 
The last build was failing due to a network outage. I reran the build and everything worked well.
